bin/circuit_transpiler/src/main.cpp

```diff
@@ -0,0 +1,495 @@
+        << ArithmetizationParams::public_input_columns << "," << ArithmetizationParams::constant_columns << ","
+        << ArithmetizationParams::selector_columns << "]," << std::endl
+        << "\"columns_rotations\":[" << std::endl;
+    for (size_t i = 0; i < columns_rotations.size(); i++) {
+        if (i != 0)
+            out << "," << std::endl;
+        out << "[";
+        for (size_t j = 0; j < columns_rotations[i].size(); j++) {
+            if (j != 0)
+                out << ",";
+            out << columns_rotations[i][j];
+        }
+        out << "]";
+    }
+    out << "]," << std::endl;
+    out << "\"r\":" << fri_params.r << "," << std::endl;
+    out << "\"step_list\":[";
+    for (size_t i = 0; i < fri_params.step_list.size(); i++) {
+        if (i != 0)
+            out << ",";
+        out << fri_params.step_list[i];
+    }
+    out << "]," << std::endl;
+    out << "\"D_omegas\":[" << std::endl;
+    for (size_t i = 0; i < fri_params.D.size(); i++) {
+        if (i != 0)
+            out << "," << std::endl;
+        out << fri_params.D[i]->get_domain_element(1).data;
+    }
+    out << "]," << std::endl;
+    out << "\"rows_amount\":" << table_description.rows_amount << "," << std::endl;
+    out << "\"max_degree\":" << fri_params.max_degree << "," << std::endl;
+    out << "\"omega\":" << fri_params.D[0]->get_domain_element(1).data << std::endl;
+    out << "}" << std::endl;
+    out.close();
+}
+
+template<typename BlueprintFieldType, typename ArithmetizationParams, typename ColumnType>
+std::tuple<std::size_t, std::size_t,
+           nil::crypto3::zk::snark::plonk_table<BlueprintFieldType, ArithmetizationParams, ColumnType>>
+    load_assignment_table(std::istream &istr) {
+    using PrivateTableType =
+        nil::crypto3::zk::snark::plonk_private_table<BlueprintFieldType, ArithmetizationParams, ColumnType>;
+    using PublicTableType =
+        nil::crypto3::zk::snark::plonk_public_table<BlueprintFieldType, ArithmetizationParams, ColumnType>;
+    using TableAssignmentType =
+        nil::crypto3::zk::snark::plonk_table<BlueprintFieldType, ArithmetizationParams, ColumnType>;
+    std::size_t usable_rows;
+    std::size_t rows_amount;
+
+    typename PrivateTableType::witnesses_container_type witness;
+    typename PublicTableType::public_input_container_type public_input;
+    typename PublicTableType::constant_container_type constant;
+    typename PublicTableType::selector_container_type selector;
+
+    istr >> usable_rows;
+    istr >> rows_amount;
+
+    for (size_t i = 0; i < witness.size(); i++) {    // witnesses.size() == ArithmetizationParams.WitnessColumns
+        ColumnType column;
+        typename BlueprintFieldType::integral_type num;
+        for (size_t j = 0; j < rows_amount; j++) {
+            istr >> num;
+            column.push_back(typename BlueprintFieldType::value_type(num));
+        }
+        witness[i] = column;
+    }
```

Comment:
This code is repeated four times with minimal changes. Could it be a function?

The comment is always about `witnesses.size`, but there's no `witnesses` variable. Instead, there's `witness` (note the difference), `public_input` and others.